src/PollingBlockTracker.test.ts

```diff
@@ -0,0 +1,3024 @@
+        async ({ provider, blockTracker }) => {
+          jest.spyOn(provider, 'sendAsync');
+
+          await blockTracker.checkForLatestBlock();
+
+          expect(provider.sendAsync).toHaveBeenCalledWith(
+            {
+              jsonrpc: '2.0' as const,
+              id: expect.any(Number),
+              method: 'eth_blockNumber' as const,
+              params: [],
+              skipCache: true,
+            },
+            expect.any(Function),
+          );
+        },
+      );
+    });
+
+    it(`should not emit the "error" event, but should throw instead if the request for the latest block number returns an error response`, async () => {
+      recordCallsToSetTimeout({ numAutomaticCalls: 1 });
+
+      await withPollingBlockTracker(
+        {
+          provider: {
+            stubs: [
+              {
+                methodName: 'eth_blockNumber',
+                response: {
+                  error: 'boom',
+                },
+              },
+            ],
+          },
+        },
+        async ({ blockTracker }) => {
+          const promiseForLatestBlock = blockTracker.checkForLatestBlock();
+          await expect(promiseForLatestBlock).rejects.toThrow(
+            'PollingBlockTracker - encountered error fetching block:\nboom',
+          );
+        },
+      );
+    });
+
+    it(`should not emit the "error" event, but should throw instead, if, while making the request for the latest block number, the provider rejects with an error`, async () => {
+      recordCallsToSetTimeout({ numAutomaticCalls: 1 });
+      const thrownError = new Error('boom');
+
+      await withPollingBlockTracker(
+        {
+          provider: {
+            stubs: [
+              {
+                methodName: 'eth_blockNumber',
+                error: 'boom',
+              },
+              {
+                methodName: 'eth_blockNumber',
+                response: {
+                  result: '0x0',
+                },
+              },
+            ],
+          },
+        },
+        async ({ blockTracker }) => {
+          const promiseForLatestBlock = blockTracker.checkForLatestBlock();
+          await expect(promiseForLatestBlock).rejects.toThrow(thrownError);
+        },
+      );
+    });
+
+    it('should update the current block number', async () => {
+      recordCallsToSetTimeout();
+
+      await withPollingBlockTracker(
+        {
+          provider: {
+            stubs: [
+              {
+                methodName: 'eth_blockNumber',
+                response: {
+                  result: '0x0',
+                },
+              },
+            ],
+          },
+        },
+        async ({ blockTracker }) => {
+          await blockTracker.checkForLatestBlock();
+          const currentBlockNumber = blockTracker.getCurrentBlock();
+          expect(currentBlockNumber).toStrictEqual('0x0');
+        },
+      );
+    });
+
+    it('should never clear the current block number later', async () => {
```

Comment:
It's not entirely clear to me what this description means. I see that the test has a shorter polling interval and block reset duration set, and one automatic `setTimeout` call. But as you've observed elsewhere, there are two `setTimeout`s set, one for block duration and one for the poll. So I'm not sure what to expect here.

Reply:
The key difference between this test and the previous test is `numAutomaticCalls: 1`, which means "if `setTimeout` is called, then go ahead and call its callback immediately". That said, 1) that's hard to spot (I initially thought they were the same test when re-reading this) and 2) that's not obvious. I've tweaked this test in [26be231](https://github.com/MetaMask/eth-block-tracker/pull/103/commits/26be231f87ff56c51ecc5bb833026960b1c594ee), hopefully that's better.

Reply:
Hmm. Honestly now I'm more confused. How does this line up with the test description? The description talks about the block number not being cleared "later", but now we never get to "later".

Reply:
So you're saying that the assertion most relevant to the description is `expect(setTimeoutRecorder.calls).toHaveLength(0);`. That because `setTimeout` was never called, we can rest assured that the cache isn't going to be cleared, so the block number is "never cleared later".

OK I think I follow what you were getting at now.

Still I don't think this is a particularly effective test, and in a sense it's not doing what the description implies. The description is saying that the current block number isn't cleared, but the test is just checking that `setTImeout` wasn't called. The relationship between `setTimeout` and the block number being cleared is an implementation detail. Relying on this makes the test fragile and hard to read.

One alternative would be for the test to actually wait for the block number to be cleared, then check to ensure it isn't. That would match the description better.

Reply:
> One alternative would be for the test to actually wait for the block number to be cleared, then check to ensure it isn't. 

Recall that we are stubbing out calls to `setTimeout`, though, so while we can wait, nothing will happen until we "replay" the captured `setTimeout` calls. Are you saying it would be more clear to not do that at all? For instance we could have:

``` typescript
      const blockResetDuration = 2000;

      await withPollingBlockTracker(
        {
          blockTracker: {
            blockResetDuration
          },
          provider: {
            stubs: [
              {
                methodName: 'eth_blockNumber',
                response: {
                  result: '0x0',
                },
              },
            ],
          },
        },
        async ({ blockTracker }) => {
          await blockTracker.checkForLatestBlock();

          await new Promise((resolve) => setTimeout(resolve, blockResetDuration));

          const currentBlockNumber = blockTracker.getCurrentBlock();
          expect(currentBlockNumber).toStrictEqual('0x0');
        },
      );
    });
```

Reply:
Yes, something like that is what I was thinking. I did fail to consider that the `setTimeout` calls were stubbed.

Is there a way for us to show that the test has gotten to "later" (i.e. that the block reset duration has been elapsed), even with the stubbing? Or was that the intent with the test as currently written?

Reply:
Hmm, I don't know if there is a way to do that with the stubbing. I wasn't trying to show that we have waited, because I didn't think we had to.

Reply:
Yes, that alternate title does match these assertions better.

Though... I don't see any problem with waiting _and_ asserting that the timeout was never set. That would show that it was "later" wouldn't it? The timer assertion covers the "we forgot to replay the catpured calls" case, and the waiting covers other methods that could be used to set the timer (`setInterval`, some other event loop or Promise queue based method, etc.)

Reply:
Okay, test updated here: [ad35cdb](https://github.com/MetaMask/eth-block-tracker/pull/103/commits/ad35cdb85356076e7943d7ec91df51df9fea38c0)